### PR TITLE
added simple configuration options

### PR DESCRIPTION
**Pull request by maltesander**
merged with: simple spark version update/downgrade 2.4 to 3.0

### CODE REVIEW

File: crd/sparkcluster.crd.yaml [hunk 1]
```diff
@@ -35,6 +35,28 @@ spec:
                             type: string
                           instances:
                             type: integer
+                          masterPort:
+                            type: integer
```
Comment (lfrancke) [inline crd/sparkcluster.crd.yaml:39]: If you want to you could add optional minimums here. It shouldn't be less than 0 or 1 I assume
some for the other ports

File: crd/src/lib.rs [hunk 2]
```diff
@@ -95,19 +95,26 @@ pub struct SparkNode {
...
     // for the selectors changes and the operator removes all nodes and rebuilds them
     #[derivative(Hash = "ignore", PartialEq = "ignore")]
     pub instances: usize,
     pub config: Option<Vec<ConfigOption>>,
     pub env: Option<Vec<ConfigOption>>,
-    // TODO: master -> use Option<T>
-    // worker -> use Option<T>
+
+    // master options
+    pub master_port: Option<usize>,
```
Comment (lfrancke) [inline crd/src/lib.rs:108]: This is a bit ... mehh.. to have master & worker stuff in the same selector (I know we talked about it in the past). But I'm fine with leaving it for now.

File: operator/src/config.rs [hunk 3]
```diff
@@ -13,59 +38,65 @@ const SPARK_URL_START: &str = "spark://";
...
-                        format!(
-                            "{}{}:{},",
-                            SPARK_URL_START, selector.node_name, config_option.value
-                        )
-                        .as_str(),
-                    );
-                    // found
-                    continue 'selectors;
-                }
+            if let Some(master) =
```
Comment (lfrancke) [inline operator/src/config.rs:52]: This is a lot of code to handle just one or two options. But I don't have a better way now either so I'm fine with keeping it as it is.

File: operator/src/lib.rs [hunk 4]
```diff
@@ -120,7 +119,7 @@ impl PodInformation {
     }
 }
 
-/// Count the pods stored in each (hashed) selector
+/// Retrieve all pods encapsulated for a spark node in selector/hash -> Vec<Pod>
```
Comment (lfrancke) [inline operator/src/lib.rs:122]: I'm not sure if `encapsulated` is the right word here. I'm not sure what you mean.

File: operator/src/config.rs [hunk 5]
```diff
@@ -125,14 +156,122 @@ pub fn create_volumes(configmap_name: &str) -> Vec<Volume> {
...
+        log_dir.to_string(),
+    );
+    if let Some(store_path) = &selector.store_path {
+        config.insert(SPARK_HISTORY_STORE_PATH.to_string(), store_path.to_string());
+    }
+    if let Some(port) = &selector.history_ui_port {
+        config.insert(SPARK_HISTORY_UI_PORT.to_string(), port.to_string());
+    }
+
+    // add config options -> may override previous settings which is what we want
```
Comment (lfrancke) [inline operator/src/config.rs:209]: Can you elaborate which "config options" you mean here?
Clarify that these are the "free form" configs or whatever we want to call them

File: operator/src/config.rs [hunk 6]
```diff
@@ -125,14 +156,122 @@ pub fn create_volumes(configmap_name: &str) -> Vec<Volume> {
...
+/// Get all required configuration options
+/// 1) from spec
+/// 2) from node
+/// 3) from selector
+/// 4) from config properties
+/// # Arguments
+/// * `spec` - SparkCluster spec for common properties
+/// * `selector` - SparkClusterSelector containing desired config properties
+///
+pub fn get_config_properties(
```
Comment (lfrancke) [inline operator/src/config.rs:180]: This seems to mix configuration for different types of nodes into one?
History Server & Worker stuff etc.?
Fine for now but in the future it'd be good if the configs are as minimal as possible.

File: operator/src/config.rs [hunk 7]
```diff
@@ -125,14 +156,122 @@ pub fn create_volumes(configmap_name: &str) -> Vec<Volume> {
...
+
+/// Get all required env variables
+/// 1) from spec
+/// 2) from node
+/// 3) from selector
+/// 4) from env variables
+/// # Arguments
+/// * `selector` - SparkClusterSelector containing desired env variables
+///
+pub fn get_env_variables(selector: &SparkNodeSelector) -> HashMap<String, String> {
```
Comment (lfrancke) [inline operator/src/config.rs:229]: Same comments as for the config_properties method

File: README.adoc [hunk 8]
```diff
@@ -12,3 +12,144 @@ It is developed against the latest Rust release (1.50.0-nightly at the time of w
 NOTE: This requires Rust nightly due to the use of the `backtrace` feature.
 
     cargo build
+
+== Configuration Options
+
+The cluster can be configured via a yaml file.
```
Comment (lfrancke) [inline README.adoc:18]:
```suggestion
The cluster can be configured via a YAML file.
```

File: README.adoc [hunk 9]
```diff
@@ -12,3 +12,144 @@ It is developed against the latest Rust release (1.50.0-nightly at the time of w
...
+[cols="1,1"]
+|===
+|Common shared options
+|Contains configuration that is shared within the whole cluster. E.g. the spark image, secrets, encryption or logging options.
+
+|Node type options
+|This configuration is shared for all nodes of a certain type (Master, Worker, History-Server)
+
+|Selector specific options
+|Options provided in the selector are limited to one certain node of a node type (Master, Worker, History-Server) and apply to one physical machine.
```
Comment (lfrancke) [inline README.adoc:60]:
```suggestion
|Options provided in the selector are limited to one specific node of a node type (Master, Worker, History-Server) and apply to one physical machine.
```

**Review by lfrancke — CHANGES_REQUESTED** [review]
Minor things only

File: crd/sparkcluster.crd.yaml (hunk 1 above)
Comment (maltesander) [reply crd/sparkcluster.crd.yaml:39]: Yes the question is: Do we catch wrong inputs already in the crd or just via product config?
For ports e.g. if we run the operators as user you cannot pick anything lower than 1024. With root you have the full range. We cannot express that in the crd as far as i know? But then why would run them as root. Just an example.

File: crd/src/lib.rs (hunk 2 above)
Comment (maltesander) [reply crd/src/lib.rs:108]: I know it looks somewhat weird but has some benefits. Less code, better to work with. Restriction are coming from CRD here. So you cannot assign the master_port for workers etc. 
For now it works, but i think we probably have to split them up at some point too.

File: operator/src/config.rs (hunk 6 above)
Comment (maltesander) [reply operator/src/config.rs:180]: With the hashed selector approach we only create one config map per selector (which may be shared for multiple master/worker...). That means e.g. a selector for a master wont have any properties from a history server.

File: operator/src/config.rs (hunk 3 above)
Comment (maltesander) [reply operator/src/config.rs:52]: I dont see another way, we have multiple options to set the ports, and i need to get the right port to connect the workers properly

File: crd/sparkcluster.crd.yaml (hunk 1 above)
Comment (lfrancke) [reply crd/sparkcluster.crd.yaml:39]: Yeah you're right but it doesn't hurt to catch things twice. And this creates some auto-generated docs.
Fine for now.

**Review by lfrancke — APPROVED** [review]

**Pull request merged**